File: Python3/1711. Count Good Meals.py
class Solution:
    def countPairs(self, deliciousness: List[int]) -> int:
        mod = 10 ** 9 + 7
        res = 0
        meals = collections.defaultdict(int)
        max_num = max(deliciousness) * 2

        for d in deliciousness:
            exp = 1
            while exp <= max_num:
                if exp - d in meals:
                    res = (res + meals[exp - d]) % mod
                exp *= 2
            meals[d] += 1
        
        return res

File: C++/1711. Count Good Meals.cpp
class Solution {
public:
    int countPairs(vector<int>& deliciousness) {
        unordered_map<int, int> meals;
        int max_num = *(max_element(deliciousness.begin(), deliciousness.end())) * 2;
        int res = 0;
        int mod = pow(10, 9) + 7;
        for(auto d:deliciousness) {
            int exp = 1;
            while(exp <= max_num) {
                if(meals.count(exp - d)) {
                    res = (res + meals[exp - d]) % mod;
                }
                exp <<= 1;
            }
            meals[d]++;
        }
        return res;
    }
};
